### PR TITLE
[Pallas TPU] Add Pipelining and BlockSpecs documentation for Pallas TPU

### PR DESCRIPTION
**Pull request by sharadmv**

### CODE REVIEW

File: docs/pallas/tpu/index.rst [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Pallas TPU
+=============================
```
Comment (jakevdp) [inline docs/pallas/tpu/index.rst:2]: nit: make underline match width of text

**Review by jakevdp — APPROVED** [review]

File: docs/pallas/tpu/pipelining.md [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+```
+This generalizes to any tuple of integers (a length `d` grid will correspond to `d` nested loops).
+
++++ {"id": "hRLr5JeyyEwM"}
+
+### `BlockSpec`, a.k.a. how to chunk up inputs
+
++++ {"id": "miWgPkytyIIa"}
+
+The next piece of information we need to provide Pallas in order to automatically pipeline our computation is information on how to chunk it up. Specifically, we need to provide a mapping between *the iteratio of the loop* to *which block of our inputs and outputs to be operated on*. A `BlockSpec` is exactly these two pieces of information.
```
Comment (yashk2810) [inline docs/pallas/tpu/pipelining.md:191]:
```suggestion
The next piece of information we need to provide Pallas in order to automatically pipeline our computation is information on how to chunk it up. Specifically, we need to provide a mapping between *the iteration of the loop* to *which block of our inputs and outputs to be operated on*. A `BlockSpec` is exactly these two pieces of information.
```

File: docs/pallas/tpu/pipelining.md [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+      out_shape=jax.ShapeDtypeStruct(out_shape, x.dtype)
+      )(x)
+sum(x)
+```
+
++++ {"id": "2828qXBI5ksZ"}
+
+This `sum` function now outputs the correct values!
+
+One last thing to note about reductions in Pallas are that **they must be done in the minormost (rightmost) dimensions of our grid**. This is because the pipeline that Pallas generates using the `BlockSpec`s, `grid` and kernel function *does not read outputs back from HBM*. Once you've written an output value back to HBM you cannot revisit it. Therefore, you cannot do a reduction across a grid dimension that has any revisiting and therefore all reductions need to happen in the rightmost dimensions.
```
Comment (yashk2810) [inline docs/pallas/tpu/pipelining.md:392]: You say to reduce on the rightmost dimension but above you are reducing on the leftmost (major) dimension right?

Maybe the example should be changed?

Comment (sharadmv) [reply docs/pallas/tpu/pipelining.md:347]: The `grid` is only one dimensional so technically we are reducing over its minormost dimension (not the minormost dimension of the inputs/outputs, it's the grid that matters). I'll add clarification.

Comment (yashk2810) [reply docs/pallas/tpu/pipelining.md]: Ahh, makes sense. Thanks!

File: docs/pallas/tpu/pipelining.md [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+  x_vregs = x_vmem_ref[...]
+  y_vregs = y_vmem_ref[...]
+  # Execute a vectorized add
+  z_vregs = x_vregs + y_vregs
+  # Store the output values in VREGs back into VMEM
+  z_vmem_ref[...] = z_vregs
+
+
+def add_matrices(x: jax.Array, y: jax.Array) -> jax.Array:
+  # pallas_call will (by default) copy `x` and `y` from HBM into VMEM. It will
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:74]: Is saying "it will also allocate VMEM buffers for input" redundant with "pallas_call will copy `x` and `y` from HBM into VMEM"?

File: docs/pallas/tpu/pipelining.md [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+Let's implement a Pallas function that does just that!
+
+```{code-cell}
+:id: 2IXQxNWrKJyb
+:outputId: d62eb493-5f92-4496-f113-d3cd24cb0b9f
+
+def add_matrices_kernel(x_vmem_ref, y_vmem_ref, z_vmem_ref):
+  # Load x and y from VMEM into VREGs
+  x_vregs = x_vmem_ref[...]
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:65]: Maybe use `[:, :]` instead of the rank-polymorphic `[...]`, given the name of the function?

File: docs/pallas/tpu/pipelining.md (hunk 4 above)
Comment (mattjj) [reply docs/pallas/tpu/pipelining.md:74]: Maybe the second sentence here meant to say "intermediates"?

File: docs/pallas/tpu/pipelining.md [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+
+## Constraints of using VMEM/SMEM
+
+Pallas exposes access to lower level memory spaces like VMEM and SMEM but writing kernels utilizing them adds some considerations.
+
+1. Memory capacity. VMEM and SMEM are *small*! VMEM on v4 TPUs is only 16MiB and SMEM ranges in the tens to hundreds of KiB. If our arrays are too big, we won't even be able to fit them into VMEM at all. For reference, a `f32[2048, 2048]` array is 16MiB, so our above kernel won't scale beyond moderately sized arrays.
+
+2. Memory bandwidth. Copying to/from HBM and VMEM takes a long time, at least compared to most compute instructions. The `add_matrices` function above will likely spend more time copying between HBM and VMEM than actually performing the addition itself.
+
+With these two constraints in mind, we'll have to rethink our strategy for getting performance of our TPUs.
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:104]:
```suggestion
With these two constraints in mind, we'll have to rethink our strategy for getting performance out of our TPUs.
```

File: docs/pallas/tpu/pipelining.md [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+
++++ {"id": "_NTqvlbetB3P"}
+
+## Primer: Pipelining
+
+Pipelining our computation offers a way of dealing with both the memory capacity and bandwidth constraints in one fell swoop. What do we mean by pipelining?
+
+The goal is: *concurrently* copy to/from HBM and VMEM *while* utilizing our compute units. However, the way we expressed our program above is that we copy *all* of `x` and `y` before we start doing any compute with them. This creates a dependence between the copy and the compute.
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:112]: Might be slightly clearer to say 'in parallel' rather than 'concurrently', since we need these things to be physically running at the same time (and overlapping) not just logically separated (e.g. https://stackoverflow.com/a/1050257).

Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:112]: Maybe put the "However, the way we expressed ..." stuff in the preceding section? ATM it feels like we're jumping back to discussing the old kernel more right when we were starting to introduce pipelining. But if we move these sentences to the preceding section, we can in that section explain what's wrong with the previous way of writing things, and in this section focus on pipelining.

File: docs/pallas/tpu/pipelining.md [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+7. Load `x2, y2` from VMEM into VREGs.
+8. Execute the `z2 = x2 + y2` using the compute units.
+9. Store `z2` into VMEM.
+10. Wait until `z1` is copied into HBM.
+10. Start copying `z2` from VMEM back into HBM.
+10. Wait until `z2` is copied into HBM.
+
+Any time we are doing compute here, we are asynchronously copying something. This means that some of the time spent copying is not wasted.
+
+The two most important numbers for determining how efficient a pipelined computation are a) how many FLOP/s we can execute and b) how many bytes/s we can copy. The ratio of these two (FLOPs/bandwidth) is called the *arithmetic intensity* of an operation.
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:134]: I don't think this is consistent with [the wikipedia definition of arithmetic intensity](https://en.wikipedia.org/wiki/Roofline_model#:~:text=Arithmetic%20intensity,-The%20arithmetic%20intensity&text=will%20be%20the%20ratio%20of,movement%20(FLOPs%2Fbyte).).  That definition is about FLOPs (ie total number of floating point operations performed, not FLOPS or FLOP/s, ie floating point operations per second) compared to total bytes of traffic (ie not bytes/sec). Also I think those quantities are properties of a workload, e.g. matrix multiplication of particular shapes, rather than a chip's FLOP/s and bytes/s. See also [this ref](https://people.eecs.berkeley.edu/~kubitron/cs252/handouts/papers/RooflineVyNoYellow.pdf).

Maybe there are some other definitions of the term 'arithmetic intensity' that are consistent with what's written here, but if so can you link them? If not, let's adapt to be consistent with Wikipedia and link that. Or just delete this paragraph because I don't think arithmetic intensity is ever referenced again in the doc.

File: docs/pallas/tpu/pipelining.md [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+
+## Primer: Pipelining
+
+Pipelining our computation offers a way of dealing with both the memory capacity and bandwidth constraints in one fell swoop. What do we mean by pipelining?
+
+The goal is: *concurrently* copy to/from HBM and VMEM *while* utilizing our compute units. However, the way we expressed our program above is that we copy *all* of `x` and `y` before we start doing any compute with them. This creates a dependence between the copy and the compute.
+
+However, if we can chunk up our computation into several subcomputations (e.g. when we add two matrices, we can express that as addition of "blocks" of the original matrices together), we can now overlap the copies of one of those subcomputations with the compute of the other. Let's walk through a simple example:
+
+Let's say we split our arrays `x` and `y` into `x1, x2` and `y1, y2` (for example, split along the leading axis, resulting in two `(256, 512)` arrays for each input. We can now execute the following pipelined computation.
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:116]: Maybe we can make a pair of timeline sketches illustrating how things look with and without pipelining for this example? (Or even better than a drawing would be if we could 'take a photo' of real timings, but I'm not sure if that's feasible.)

File: docs/pallas/tpu/pipelining.md [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+The number of iterations in the loop is specified via the `grid` argument to `pallas_call`. Conceptually:
+```python
+pl.pallas_call(some_kernel, grid=n)(...)
+```
+maps to
+```python
+for i in range(n):
+  # do HBM<->VMEM copies
+  some_kernel(...)
+  # do HBM<->VMEM copies
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:159]: Should the first one be `HBM -> VMEM` and the second one be `VMEM -> HBM`?

File: docs/pallas/tpu/pipelining.md [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+x_block_spec = pl.BlockSpec(x_index_map, (256, 512))
+```
+
+The `y_block_spec` and `z_block_spec` will be identical to `x_block_spec`.
+
++++ {"id": "noybOKghzjwG"}
+
+### Putting it together
+
+We provide these arguments to `pallas_call` via `grid`, `in_specs` and `out_specs` (`in_specs` correspond to inputs, `out_spec` correspond to outputs).
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:204]:
```suggestion
We provide these arguments to `pallas_call` via `grid`, `in_specs` and `out_specs` (`in_specs` corresponds to the tuple of positional arguments, and `out_spec` corresponds to the output).
```

File: docs/pallas/tpu/pipelining.md [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+
++++ {"id": "7Xtz9oMs0ZRL"}
+
+### Parameterizing a pipeline
+
++++ {"id": "esY4GcIB0bqQ"}
+
+It's common to parameterize the block shapes in our kernel. Block sizes are perhaps the most important parameter to tune when optimizing the performance of Pallas kernels! They us more control over the pipeline (for example, picking smaller blocks adds more iterations to our pipelined loop where each iteration has less work to do).
+
+Furthermore, we could also carve up the inputs and outputs along the 2nd dimension (we are only splitting along the first right now). Let's write more general kernel that handles both of these additions.
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:236]:
```suggestion
Furthermore, we could also carve up the inputs and outputs along the 2nd dimension (we are only splitting along the first right now). Let's write a more general kernel that handles both of these additions.
```

File: docs/pallas/tpu/pipelining.md [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+
+One last thing to note about reductions in Pallas are that **they must be done in the minormost (rightmost) dimensions of our grid** (our grid is 1-dimensional in the above example so we are reducing over its minormost dimension). This is because the pipeline that Pallas generates using the `BlockSpec`s, `grid` and kernel function *does not read outputs back from HBM*. Once you've written an output value back to HBM you cannot revisit it. Therefore, you cannot do a reduction across a grid dimension that has any revisiting and therefore all reductions need to happen in the rightmost dimensions.
+
++++ {"id": "KvPFez9N8cKJ"}
+
+## TPUs in Megacore configuration
+
++++ {"id": "0f4HAVzQ8n71"}
+
+Some TPU chips have two TensorCores but appear as one device to JAX users. This is what is called "megacore". The separate TensorCores have their own separate VMEM, VREGs, SMEM, SREGs and compute units but *share HBM*.
```
Comment (mattjj) [inline docs/pallas/tpu/pipelining.md:355]:
```suggestion
Some TPU chips have two TensorCores but appear as one device to JAX users. This is called "megacore". The separate TensorCores have their own separate VMEM, VREGs, SMEM, SREGs and compute units but *share HBM*.
```

File: docs/pallas/tpu/pipelining.md (hunk 7 above)
Comment (sharadmv) [reply docs/pallas/tpu/pipelining.md:112]: This one is tough to move up because I'm trying to explain why we can't naively pipeline our code above because of the dependence, to motivate chunking. I reworded it a bit but I'll keep it here.

File: docs/pallas/tpu/pipelining.md (hunk 8 above)
Comment (sharadmv) [reply docs/pallas/tpu/pipelining.md:134]: Fixed.

File: docs/pallas/tpu/pipelining.md (hunk 9 above)
Comment (sharadmv) [reply docs/pallas/tpu/pipelining.md:116]: Anselm pointed me to how to make animations of this, so I'll try that and add it as a follow up!
